src/main/java/net/neoforged/neoforge/capabilities/BlockCapabilityCache.java

```diff
@@ -0,0 +1,120 @@
+/*
+ * Copyright (c) NeoForged and contributors
+ * SPDX-License-Identifier: LGPL-2.1-only
+ */
+
+package net.neoforged.neoforge.capabilities;
+
+import java.util.Objects;
+import java.util.function.BooleanSupplier;
+import net.minecraft.core.BlockPos;
+import net.minecraft.server.level.ServerLevel;
+import net.minecraft.world.level.Level;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * A cache for block capabilities, to be used to track capabilities at a specific position, with a specific context.
+ * <p>
+ * The cache is invalidated when the level is notified of a change via {@link Level#invalidateCapabilities(BlockPos)}.
+ * <p>
+ * Instances are automatically cleared by the garbage collector when they are no longer in use.
+ */
+public final class BlockCapabilityCache<T, C> {
+    /**
+     * Creates a new cache instance and register it to the level.
+     *
+     * @param capability the capability
+     * @param level      the level
+     * @param pos        the position
+     * @param context    extra context for the query
+     */
+    public static <T, C> BlockCapabilityCache<T, C> create(BlockCapability<T, C> capability, ServerLevel level, BlockPos pos, C context) {
+        return create(capability, level, pos, context, () -> true, () -> {});
```

Comment:
This is somewhat a pre-optimization, but we should store the last two parameters as private static instances so that if people actually make use of the system we don't potentially create a bunch of tiny extra objects for now reason.
```java
private static final BooleanSupplier ALWAYS_VALID = () -> true;
private static final Runnable NO_INVALIDATION = () -> {
};
```

Reply:
That won't happen because these lambdas are non-allocating